### ISSUE TITLE
http: Unmarshalling implementation makes it hard to propagate exceptions to a rejection handler

### ISSUE THREAD

**Issue by stanch** [akka/akka#19681]
Please see a more elaborate explanation here: https://github.com/hseeberger/akka-http-json/issues/46, /cc @hseeberger.

TL;DR: [This code](https://github.com/akka/akka/blob/master/akka-http/src/main/scala/akka/http/scaladsl/server/directives/MarshallingDirectives.scala#L37-L38):
```scala
case Failure(x: IllegalArgumentException) ⇒ reject(ValidationRejection(x.getMessage.nullAsEmpty, Some(x)))
case Failure(x) ⇒ reject(MalformedRequestContentRejection(x.getMessage.nullAsEmpty, Option(x.getCause)))
```
means that if an exception thrown in an unmarshaller needs to be available in the rejection handler (e.g. the exception is a case class with useful semantic information), the said exception needs to either extend `IllegalArgumentException`, or needs to be wrapped in another one as a cause.

Is there any specific reason why the second case does not use `Option(x)` instead of `Option(x.getCause)`?

**Comment by johanandren** [comment]
From what I can see is that it has been like that since the initial port from spray. The `ValidationRejection` case was added in #15792.

@jrudolph @sirthias could either of you shed any historical light on this?

**Comment by sirthias** [comment]
Thanks, @stanch, for the excellent writeup of the issue in this ticket and the underlying one hseeberger/akka-http-json#46 !

IIUC you'd like to tunnel a rejection through an exception.
We already have a vehicle for this on the server-side: `akka.http.scaladsl.server.RejectionError`,
which for some reason doesn't have its own case branch in the match you already pointed to.
AFAICS this is an oversight that we should fix.

Specifically, if we add

```scala
case Failure(RejectionError(r)) ⇒ reject(r)
```

we (and you) should be all set, right?

**Comment by stanch** [comment]
@sirthias Actually I think I would like the opposite :) That is, to tunnel an exception through a rejection. In the example I gave in the linked issue, the exception that occurs during JSON mapping has information on the errors encountered. I’d like to report those errors to the client, but when I receive the exception in the rejection handler, it’s only in the `message` of `MalformedRequestContentRejection` — I don’t have access to the underlying exception that has all the details. IMHO it would make sense for it to be in the cause, but the code above expects the exception to contain some other exception as its cause, which in my case it doesn’t.

**Comment by sirthias** [comment]
@stanch I see.

So here's what I'd propose:

1. We change `MalformedRequestContentRejection::cause` member to `Throwable` rather than an `Option[Throwable]`.
2. In the snippet you quoted above in the ticket description we pack the `x` itself, rather than the `x.cause` into the `MalformedRequestContentRejection`.
3. We add the `case Failure(RejectionError(r)) ⇒ reject(r)` case.

IIUC this should fix all the issues that came up over the course of this ticket discussion, correct?

**Comment by jrudolph** [comment]
:+1:

**Comment by stanch** [comment]
Sounds great!

**Comment by hseeberger** [comment]
Any update on this?

**Comment by ktoso** [comment]
Went under the radar so far, I'm triaging tickets in bunches of 50 daily but didn't get here yet. 
Seems we have a way forward, just a matter of actually implementing it.

**Comment by stanch** [comment]
Just tested and it is solved in `2.4.5`, great!

**Comment by ktoso** [comment]
Thanks for testing, cheers!
